src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp

```diff
@@ -484,6 +484,9 @@ static void do_primitives() {
 static void do_unloading_klass(Klass* klass) {
   assert(klass != nullptr, "invariant");
   assert(_subsystem_callback != nullptr, "invariant");
+  if (klass->is_instance_klass() && InstanceKlass::cast(klass)->is_scratch_class()) {
```

Comment:
This is not correct. The unloaded event is generated earlier, this part is the metadata system serializing the constants referenced in said event. The result would still be an event, but it would have missing constants, such as the klass name.

Can scratch klasses short-circuit the "external" unload paths in cases of scratch klasses (which are really only implementation details). There is a property in the classfile parser about "broadcasting" that we (jfr) use to only log a single IK instance when we replace an IK with another schema extended IK.

Logically there should only be one official instance of an IK.

Reply:
Thanks for suggestion, I filed https://bugs.openjdk.org/browse/JDK-8341421